Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs

```diff
@@ -218,53 +217,78 @@ static FieldDescriptor GetField(MessageDescriptor descriptor, string name)
                 return field;
             }
 
-            static Regex ParsePattern(string pattern)
+            static (Regex, Func<Match, string>) ParsePattern(string pattern)
             {
                 var builder = new StringBuilder("^"); // We want to match the whole string
+                Action<Match, StringBuilder> formatBuilder = null;
                 int currentIndex = 0;
+                int groupIndex = 1;
                 while (currentIndex < pattern.Length)
                 {
                     int starStart = pattern.IndexOf('*', currentIndex);
+                    // Handle the literal between the previous star (if any) and the start of the next one.
+                    // If starStart==-1, this is any trailing literal at the end of the pattern.
+                    string literal = pattern.Substring(currentIndex, (starStart == -1 ? pattern.Length : starStart) - currentIndex);
+                    if (literal != "")
+                    {
+                        builder.Append(Regex.Escape(literal));
+                        formatBuilder += (match, sb) => sb.Append(literal);
+                    }
                     if (starStart < 0)
                     {
                         break;
                     }
-                    builder.Append(Regex.Escape(pattern.Substring(currentIndex, starStart - currentIndex)));
+
                     int starEnd = starStart + 1;
                     while (starEnd < pattern.Length && pattern[starEnd] == '*')
                     {
                         starEnd++;
                     }
                     int starCount = starEnd - starStart;
+                    // Deliberately local to have a different variable per 
+                    int matchGroupIndex = groupIndex;
+                    groupIndex++;
+                    // We match any characters within the regex, regardless of number of stars. It just affects
+                    // how the value is escaped. This uses a reluctant matcher, which means x/*/** (which is more likely
+                    // than x/**/*) will match x/a/b/c as "*=a" and then "**=b/c" which is probably the best result.
```

Comment:
Could /**/* mean that there needs to be at least one segment at the end? Definetely something that should be in the spec. Or maybe even something like /**/*/** meaning there should be at least one segment?

Reply:
(I've updated the comment as the stars with the slashes were being used for bold formatting, so I've now put them inside code blocks)

Reply:
Those would be valid patterns, yes - but `*/**` would accomplish the same thing and be more "usual". Fundamentally it's a bit ambiguous here.